### PR TITLE
feat: playerresize event in all cases

### PR DESCRIPTION
**Pull request by gkatsev**
Use [ResizeObserver](https://github.com/WICG/ResizeObserver) when available for better and more performant resizing information, otherwise, fall back to a throttled `resize` event on an iframe that's the size of the player.
Allows a video.js user to disable this by setting `resizeManager: false` as an option since the component will not be initialized.
This probably needs to revert #4800 (https://github.com/videojs/video.js/commit/e0ed0b5cd762f83dc54a7b56e8774718485e924d) because we end up getting two `playerresize` events with the dimension methods now.

TODO:
* [x] Revert #4800 
* [x] fix current tests
* [x] make RO polyfill possible
* [x] allow RO to be disabled
* [x] debounce calls instead of throttle
* [x] Write tests
* [x] documentation
* [x] make sure things are disposed properly
  * [x] events
  * [x] objects

### CODE REVIEW

**Comment by gkatsev** [comment]
One thing to consider as well here is that we're going to get multiple `playerresize` events when changing into fullscreen, depending on what the throttle is, it can be anywhere from 1 to 3 events. Currently, the ResizeObserver isn't throttled but the iframe `resize` event is throttled to 50ms. It's probable that we want to debounce rather than throttle.

File: test/unit/resize-manager.test.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+      }
+    }
+
+    let playerresizeCalled = 0;
+    const rm = new ResizeManager(this.player, {ResizeObserver: MyResizeObserver});
+
+    this.player.on('playerresize', function() {
+      playerresizeCalled++;
+    });
+    rm.resizeObserver_.observer();
```
Comment (ldayananda) [inline test/unit/resize-manager.test.js]: `observe`?

Comment (ldayananda) [reply test/unit/resize-manager.test.js]: actually looks like you don't need this line at all

File: test/unit/resize-manager.test.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    this.clock.tick(100);
+
+    assert.equal(playerresizeCalled, 1, 'playerresize was triggered');
+
+    rm.dispose();
+  });
```
Comment (ldayananda) [inline test/unit/resize-manager.test.js]: one good test to add would be a test that makes sure a resizeObserver is created when one is available on the window

File: src/js/player.js [hunk 3]
```diff
@@ -730,17 +720,6 @@ class Player extends Component {
...
-
-    // skipListeners allows us to avoid triggering the resize event when setting both width and height
-    if (this.isReady_ && !skipListeners) {
-      /**
-       * Triggered when the player is resized.
-       *
-       * @event Player#playerresize
-       * @type {EventTarget~Event}
-       */
-      this.trigger('playerresize');
```
[the remark this answers is not in the document]
Comment (gkatsev) [reply src/js/player.js]: We are falling back to using an iframe for playerresize events when ResizeObserver isn't available, currently, that's basically everywhere, though, chrome will be shipping ResizeObserver soon.

File: src/js/resize-manager.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import * as Events from './utils/events.js';
+import mergeOptions from './utils/merge-options.js';
+import Component from './component.js';
+
+class ResizeManager extends Component {
+  constructor(player, options) {
+    let RESIZE_OBSERVER_AVAILABLE = options.ResizeObserver || window.ResizeObserver;
+
+    if (options.ResizeObserver === null) {
+      RESIZE_OBSERVER_AVAILABLE = false;
```
[the remark this answers is not in the document]
Comment (gkatsev) [reply src/js/resize-manager.js]: Yup, the idea is if you wanted to disable the ResizeObserver for whatever reason you could pass in `null`.

File: test/unit/resize-manager.test.js (hunk 2 above)
Comment (gkatsev) [reply test/unit/resize-manager.test.js]: It would be nice but mocking out `window` is a pain :)

File: test/unit/resize-manager.test.js (hunk 1 above)
Comment (gkatsev) [reply test/unit/resize-manager.test.js]: I set this in the constructor of MyResizeObserver, calling it calls the observe handler which will trigger `playerresize` which we can then test. Maybe I should just make it a local variable to be more obvious what is going on.

File: src/js/resize-manager.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (ldayananda) [reply src/js/resize-manager.js]: Gotcha. I'd suggest a comment over this block so it's clear for future readers 😄

File: test/unit/resize-manager.test.js (hunk 1 above)
Comment (ldayananda) [reply test/unit/resize-manager.test.js]: Ah, I see. I agree that making it a local variable will make it more clear

File: src/js/resize-manager.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+      this.on('load', this.loadListener_);
+    }
+  }
+
+  createEl() {
+    return super.createEl('iframe', {
+      className: 'vjs-resize-manager'
+    });
+  }
```
Comment (misteroneill) [inline src/js/resize-manager.js]: I believe the element will be created even when it's not needed. Should we suppress the creation of it when `ResizeObserver` is available?

**Comment by pagarwal123** [comment]
QA-LGTM

**Pull request merged**